contracts/DutchAuction.sol

```diff
@@ -0,0 +1,176 @@
+import "erc721a/contracts/ERC721A.sol";
+
+import "./IDutchAuction.sol";
+import "./ERC721M.sol";
+
+contract DutchAuction is IDutchAuction, ERC721M, ReentrancyGuard {
+    bool private immutable _refundable;
+    uint256 private _settledPriceInWei;
+    Config private _config;
+    mapping(address => User) private _userData;
+
+    constructor(
+        // ERC721M
+        string memory collectionName,
+        string memory collectionSymbol,
+        string memory tokenURISuffix,
+        uint256 maxMintableSupply,
+        uint256 globalWalletLimit,
+        address cosigner,
+        bool refundable
+    )
+        ERC721M(
+            collectionName,
+            collectionSymbol,
+            tokenURISuffix,
+            maxMintableSupply,
+            globalWalletLimit,
+            cosigner
+        )
+    {
+        _refundable = refundable;
+    }
+
+    modifier validTime() {
+        Config memory config = _config;
+        if (
+            block.timestamp > config.endTime ||
+            block.timestamp < config.startTime
+        ) revert InvalidStartEndTime();
+        _;
+    }
+
+    function setConfig(
+        uint256 startAmountInWei,
+        uint256 endAmountInWei,
+        uint256 startTime,
+        uint256 endTime,
+        bool roundUp
+    ) external onlyOwner {
+        if (startTime == 0 || startTime >= endTime)
+            revert InvalidStartEndTime();
+        if (startAmountInWei == 0) revert InvalidAmountInWei();
+
+        Config memory config;
+        config = Config({
+            startAmountInWei: startAmountInWei,
+            endAmountInWei: endAmountInWei,
+            startTime: startTime,
+            endTime: endTime,
+            roundUp: roundUp
+        });
+        _config = config;
+    }
+
+    function getConfig() external view returns (Config memory) {
+        return _config;
+    }
+
+    function getSettledPriceInWei() external view returns (uint256) {
+        return _settledPriceInWei;
+    }
+
+    function getCurrentPriceInWei() public view returns (uint256) {
+        Config memory config = _config; // storage to memory
+        uint256 amount;
+        bool roundUp = config.roundUp;
+
+        if (config.startAmountInWei != config.endAmountInWei) {
+            // Declare variables to derive in the subsequent unchecked scope.
+            uint256 duration;
+            uint256 elapsed;
+            uint256 remaining;
+
+            // Skip underflow checks as startTime <= block.timestamp < endTime.
+            unchecked {
+                // Derive the duration for the order and place it on the stack.
+                duration = config.endTime - config.startTime;
+
+                // Derive time elapsed since the order started & place on stack.
+                elapsed = block.timestamp - config.startTime;
+
+                // Derive time remaining until order expires and place on stack.
+                remaining = duration - elapsed;
```

Comment:
Remaining will be underflow because `elapsed` could be > `duration`.

We should either:
1. check `validTime` to call this fn
2. return `startAmountInWei` if before start time, and return `endAmountInWei` if after end time

wdyt?

Reply:
prefer option 2, so the UI can correctly display time even after the sale has ended

Reply:
+1. Updated.